pageserver/src/tenant/timeline.rs

```diff
@@ -3395,8 +3395,10 @@ impl Timeline {
                 return Err(GetVectoredError::Cancelled);
             }
 
-            let keys_done_last_step = reconstruct_state.consume_done_keys();
+            let (keys_done_last_step, image_coverage_last_step) =
+                reconstruct_state.consume_done_keys();
             unmapped_keyspace.remove_overlapping_with(&keys_done_last_step);
+            unmapped_keyspace.remove_overlapping_with(&image_coverage_last_step);
```

Comment:
I think we need to make `get_vectored_reconstruct_data_timeline` return two keyspace: the one it has completed and the one that it know for sure won't complete and then track that in `get_vectored_reconstruct_data`. Something like this:
```
diff --git a/pageserver/src/tenant/timeline.rs b/pageserver/src/tenant/timeline.rs
index d2fcd6c4a..4e6d45123 100644
--- a/pageserver/src/tenant/timeline.rs
+++ b/pageserver/src/tenant/timeline.rs
@@ -3291,13 +3291,14 @@ impl Timeline {
         let mut timeline = self;
 
         let mut cont_lsn = Lsn(request_lsn.0 + 1);
+        let missing_keyspace = KeySpace::default();
 
         loop {
             if self.cancel.is_cancelled() {
                 return Err(GetVectoredError::Cancelled);
             }
 
-            let completed = Self::get_vectored_reconstruct_data_timeline(
+            let (mut completed, missing) = Self::get_vectored_reconstruct_data_timeline(
                 timeline,
                 keyspace.clone(),
                 cont_lsn,
@@ -3307,7 +3308,12 @@ impl Timeline {
             )
             .await?;
 
-            keyspace.remove_overlapping_with(&completed);
+            if !missing.is_empty() {
+                missing_keyspace.merge(&missing);
+            }
+
+            let to_remove = completed.merge(&missing);
+            keyspace.remove_overlapping_with(&to_remove);
 
             // Do not descend into the ancestor timeline for aux files.
             // We don't return a blanket [`GetVectoredError::MissingKey`] to avoid
@@ -3331,12 +3337,24 @@ impl Timeline {
             timeline = &*timeline_owned;
         }
 
-        if keyspace.total_raw_size() != 0 {
+        let missing = {
+            if !keyspace.is_empty() {
+                Some(keyspace)
+            } else if !missing_keyspace.is_empty() {
+                Some(missing_keyspace)
+            } else {
+                None
+            }
+        };
+
+
+
+        if let Some(missing_keyspace) = missing {
             return Err(GetVectoredError::MissingKey(MissingKeyError {
-                key: keyspace.start().unwrap(), /* better if we can store the full keyspace */
+                key: missing_keyspace.start().unwrap(), /* better if we can store the full keyspace */
                 shard: self
                     .shard_identity
-                    .get_shard_number(&keyspace.start().unwrap()),
+                    .get_shard_number(&missing_keyspace.start().unwrap()),
                 cont_lsn,
                 request_lsn,
                 ancestor_lsn: Some(timeline.ancestor_lsn),
```

Reply:
unfortunately there's no way of merging two overlapping keyspaces, so I return a `Vec<KeySpace>` for now. Also moved the code around to ensure aux keyspace still works.

Reply:
... and I don't think we need to maintain `missing_keyspace` outside the loop?